### ISSUE TITLE
host-char mismatches with the URL Standard

### ISSUE THREAD

**Issue by annevk** [w3c/webappsec-csp#592]
E.g., CSP cannot match hosts that have an `_`, which are known to exist in the wild.

It seems rather dubious that CSP contains its own mini incompatible URL parser with extensions. And now there's also https://wicg.github.io/urlpattern/. Has anyone looked into reconciliation?

**Comment by arichiv** [comment]
@wanderview PTAL

**Comment by arichiv** [comment]
[RFC 1035](https://www.rfc-editor.org/rfc/rfc1035) prevents _ but it seems like the [url spec](https://url.spec.whatwg.org/#host-miscellaneous) allows it. In this case we could re-write the spec to delegate URL parsing to the spec and avoid refining it, but we have to be careful about our support for wildcards since the URL spec doesn't have those. I don't think this is high priority.

**Comment by clelland** [comment]
Yeah, not all domans are DNS names, so we should be following URL here. (Despite the reference there, which is probably irrelevant)

[RFC 3986](https://www.rfc-editor.org/rfc/rfc3986) and the URL spec agree that both `_` and `*` are allowed in a host, even if `*` isn't defined to be a wildcard there.

**Comment by clelland** [comment]
[RFC 1035](https://www.ietf.org/rfc/rfc1035.txt#:~:text=labels%20can%20contain%20any%208%20bit%20values) doesn't actually prevent anything, although it strongly recommends not using anything other than alphanumeric characters and hyphens. We're in a weird situation of having `*` allowed by all definitions of a host, even if it isn't used in practice, but also being treated in CSP as a special character.

We should do something to handle it; excluding it from host-char probably makes sense (while including the rest of the non-[forbidden-domain-code-point](https://url.spec.whatwg.org/#forbidden-domain-code-point)s if we can). It means that a domain in the wild which contains a literal `*` cannot be spelled out in CSP; it can only match a host pattern if it is covered by a wildcard host segment. (This is the current situation)

So the URL `https://breaking*things.example.com/` still matches `https://*.example.com` or `*`, even if you can't specify the exact hostname in a CSP declaration.

I expect that the alternative would be to define some kind of escaping convention in CSP, but that could introduce new problems with existing policies.

**Comment by arichiv** [comment]
I see, so you're saying we should support * in origins but not in CSP-sources (matching targets)?
